### PR TITLE
miRBind model training

### PR DESCRIPTION
**Pull request by evaklimentova**
Scripts to support the retraining of miRBind CNN on different subsets of Manakov 1:1 train set

### CODE REVIEW

File: code/model_training/README.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+## 2) Data encoding
+
+Run the [encode dataset](encode_dataset.py) script to transform dataset in tsv into numpy miRNA x target binding matrix.
+
+`python encode_dataset.py -i <input_dataset> -o <output_prefix>`
+
+## 3) Train the model
+
+Have a look at the [training](training.py) script with model training.
```
Comment (katarinagresova) [inline code/model_training/README.md:27]: I would write a description what this script does and what are the parameters

File: code/model_training/README.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
+# miRBind CNN model retraining
```
Comment (katarinagresova) [inline code/model_training/README.md:1]: Can we add a bash script or something showing how were these scripts executed to train models shown in the paper?

File: code/model_training/encode_dataset.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
+import pandas as pd
+import numpy as np
+import argparse
+import time
+
+
+def binding_encoding(df, alphabet, tensor_dim=(50, 20, 1)):
```
Comment (katarinagresova) [inline code/model_training/encode_dataset.py:7]: Could there be some default alphabet here? So I can use this method without worrying I will select a different alphabet

File: code/model_training/encode_dataset.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    """
+    Transform input sequence pairs to a binding matrix with corresponding labels.
+
+    Parameters:
+    - df: Pandas DataFrame with columns "noncodingRNA", "gene", "label"
```
Comment (katarinagresova) [inline code/model_training/encode_dataset.py:12]: Can we make names of columns parameters? Default values would be the one mentioned here. But I could use this method even on data with different names of columns

File: code/model_training/encode_dataset.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    """
+    labels = df["label"].to_numpy()
+
+    # Initialize dot matrix with zeros
+    ohe_matrix_2d = np.zeros((len(df), *tensor_dim), dtype="float32")
+
+    df = df.reset_index(drop=True)
+
+    # Compile matrix with Watson-Crick interactions
+    for index, row in df.iterrows():
```
Comment (katarinagresova) [inline code/model_training/encode_dataset.py:27]: Not sure if it is possible, or if it would help, but had you tried using df.apply()?

File: code/model_training/encode_dataset.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+                    break
+                base_pairs = bind_nt + ncrna_nt
+                ohe_matrix_2d[index, bind_index, ncrna_index, 0] = alphabet.get(base_pairs, 0)
+
+    return ohe_matrix_2d, labels
+
+
+def encode_large_tsv_to_numpy(tsv_file_path, data_output_path, labels_output_path, chunk_size=10000):
+    """
+    Encode a large TSV file into a NumPy matrix using chunk processing.
```
Comment (katarinagresova) [inline code/model_training/encode_dataset.py:40]: Maybe mention that nothing is being returned. Results are being stored into a file

File: code/model_training/encode_dataset.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+    # Get total number of rows in the dataset
+    num_rows = sum(len(df) for df in pd.read_csv(tsv_file_path, sep='\t', usecols=[0], chunksize=chunk_size))
+
+    # Determine the shape of the output arrays
+    labels_shape = (num_rows,)
+    data_shape = (num_rows, *tensor_dim)
+
+    # Create memory-mapped files
+    ohe_matrix_2d = np.memmap(data_output_path, dtype='float32', mode='w+', shape=data_shape)
```
Comment (katarinagresova) [inline code/model_training/encode_dataset.py:60]: I don't know these memory mapped files. How do they behave during exception in a code? Will they close the file? Or should we put it into try-catch block and close the file when something happens?

File: code/model_training/encode_dataset.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+    labels = np.memmap(labels_output_path, dtype='float32', mode='w+', shape=labels_shape)
+
+    row_offset = 0
+
+    # Process each chunk
+    for chunk in pd.read_csv(tsv_file_path, sep='\t', chunksize=chunk_size):
+        encoded_data, encoded_labels = binding_encoding(chunk, alphabet, tensor_dim)
+
+        # Write the chunk's data and labels to the memory-mapped files
+        ohe_matrix_2d[row_offset:row_offset + len(chunk)] = encoded_data
```
Comment (katarinagresova) [inline code/model_training/encode_dataset.py:70]: Is this more effective compared to appending to a file?

File: code/model_training/training.py [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+import matplotlib.pyplot as plt
+from tensorflow.keras.layers import (
+                                BatchNormalization, LeakyReLU,
+                                Input, Dense, Conv2D,
+                                MaxPooling2D, Flatten, Dropout)
+from tensorflow.keras.optimizers import Adam
+from tensorflow.keras.utils import Sequence
+
+
+def make_architecture():
```
Comment (katarinagresova) [inline code/model_training/training.py:18]: Can you add a link to original implementation of this model?

File: code/model_training/training.py [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+            np.random.shuffle(self.indices)
+
+
+def train_model(data, labels, dataset_size, ratio, model_file, debug=False):
+
+    # set random state for reproducibility
+    np.random.seed(42)
+    tf.random.set_seed(42)
+    K.utils.set_random_seed(42)
+    # TODO still not fully reproducible? why?
```
Comment (katarinagresova) [inline code/model_training/training.py:165]: Maybe we can look into deterministic execution on GPU? https://stackoverflow.com/a/60911053
Shuffling of samples should be done in the same way every time, right?

File: code/model_training/training.py [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+
+    train_data_gen = DataGenerator(data, labels, dataset_size, batch_size=32, ratio=ratio, validation_split=0.1, is_validation=False)
```
Comment (katarinagresova) [inline code/model_training/training.py:167]: When you call it 2x here, aren't there `self.data` and `self.labels` twice?

File: code/model_training/README.md (hunk 2 above)
Comment (evaklimentova) [reply code/model_training/README.md:1]: I created a bash script to run the whole retraining "analysis"

File: code/model_training/README.md (hunk 1 above)
Comment (evaklimentova) [reply code/model_training/README.md]: Some description added

File: code/model_training/README.md (hunk 2 above)
Comment (stephaniesamm) [reply code/model_training/README.md:1]: From the 'pipeline' it seems that for subsetting, since, for example for the smallest subset, we pass `--N 200` and this is used to select N number of pos and N number of negative, then we have a total subset size of 400 for the smallest right? This is inconsistent with what we write in the paper. 

But then later for training, we pass `--dataset_size 200` to load the numpy files.. 

So I am wondering what is the actual size of the smallest subset for example, 200 or 400?

Also, if we are going to make changes to the code, I suggest we add the ratio / number of negatives as an argument to the make_subset.py file, and to consider the ratio in the DataGenerator class in the training.py file (it only works with 1:1 ratio as it is now, if I am not mistaken). I was already starting to work on some version of this. Let me know if I should continue and push something, or if you are working on this please :) Thanks!

**Comment by stephaniesamm** [comment]
From the 'pipeline' it seems that for subsetting, since, for example for the smallest subset, we pass `--N 200` and this is used to select N number of pos and N number of negative, then we have a total subset size of 400 for the smallest right? This is inconsistent with what we write in the paper. 

But then later for training, we pass `--dataset_size 200` to load the numpy files.. 

So I am wondering what is the actual size of the smallest subset for example, 200 or 400?

Also, if we are going to make changes to the code, I suggest we add the ratio / number of negatives as an argument to the make_subset.py file, and to consider the ratio in the DataGenerator class in the training.py file (it only works with 1:1 ratio as it is now, if I am not mistaken). I was already starting to work on some version of this. Let me know if I should continue and push something, or if you are working on this please :) Thanks!

File: code/model_training/encode_dataset.py [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+    df = df.reset_index(drop=True)
+
+    # Compile matrix with Watson-Crick interactions
+    for index, row in df.iterrows():
+        for bind_index, bind_nt in enumerate(row['gene'].upper()):
+            for ncrna_index, ncrna_nt in enumerate(row['noncodingRNA'].upper()):
+                if ncrna_index >= tensor_dim[1]:
+                    break
+                base_pairs = bind_nt + ncrna_nt
+                ohe_matrix_2d[index, bind_index, ncrna_index, 0] = alphabet.get(base_pairs, 0)
```
[the remark this answers is not in the document]
Comment (evaklimentova) [reply code/model_training/encode_dataset.py:37]: added a similar `if` statement for gene sequence
when the sequence is longer, the script just considers the first 20/50 nucleotides
I kept the `enumerate()` thing so it handles well shorter sequences (it just keeps part of the matrix empty)

File: code/model_training/encode_dataset.py (hunk 7 above)
Comment (evaklimentova) [reply code/model_training/encode_dataset.py]: I didn't know it before either, but it sounded like a nice concept. 
You're right with the exception, it stays open, so it's not a nice behaviour. I added the try-catch

File: code/model_training/encode_dataset.py (hunk 8 above)
Comment (evaklimentova) [reply code/model_training/encode_dataset.py]: I'm not sure, not tested, but I think if not faster, then it should be at least very similar.

File: code/model_training/encode_dataset.py (hunk 5 above)
Comment (evaklimentova) [reply code/model_training/encode_dataset.py:29]: You're right. It's not a super speed-up, but it helps a bit -> rewritten.

**Comment by evaklimentova** [comment]
> @evaklimentova From the 'pipeline' it seems that for subsetting, since, for example for the smallest subset, we pass `--N 200` and this is used to select N number of pos and N number of negative, then we have a total subset size of 400 for the smallest right? This is inconsistent with what we write in the paper.
> 
> But then later for training, we pass `--dataset_size 200` to load the numpy files..
> 
> So I am wondering what is the actual size of the smallest subset for example, 200 or 400?
> 
> Also, if we are going to make changes to the code, I suggest we add the ratio / number of negatives as an argument to the make_subset.py file, and to consider the ratio in the DataGenerator class in the training.py file (it only works with 1:1 ratio as it is now, if I am not mistaken). I was already starting to work on some version of this. Let me know if I should continue and push something, or if you are working on this please :) Thanks!

Regarding the N parameter - you're completely right and I'm just stupid and I wrote it the bad way... Fixed now.

File: code/model_training/model_training_analysis.sh [hunk 13]
```diff
@@ -0,0 +1,43 @@
+#!/bin/bash
+
+
+# Parse command-line input
+while getopts "i:" opt; do
+  case $opt in
+    i) input_file="$OPTARG" ;;
```
[the remark this answers is not in the document]
Comment (evaklimentova) [reply code/model_training/model_training_analysis.sh]: Added

File: code/model_training/model_training_analysis.sh [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+    *) echo "Usage: $0 -i <input_file>" >&2; exit 1 ;;
```
[the remark this answers is not in the document]
Comment (evaklimentova) [reply code/model_training/model_training_analysis.sh]: Both arguments added

File: code/model_training/model_training_analysis.sh [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+# Check if input_file is set, otherwise print usage and exit
+if [ -z "$input_file" ]; then
+  echo "Error: -i <input_file> is mandatory"
+  echo "Usage: $0 -i <input_file>"
+  exit 1
+fi
+
+# Make subsets of the dataset
+echo "Creating dataset subsets..."
+python make_subset.py --N 200 --dataset "$input_file" --output Manakov_subset_200.tsv
```
[the remark this answers is not in the document]
Comment (evaklimentova) [reply code/model_training/model_training_analysis.sh]: Code rewritten

File: code/model_training/training.py (hunk 11 above)
Comment (evaklimentova) [reply code/model_training/training.py]: Yes, they are. But I'm not sure how to solve it nicely. On one hand, having the data twice shouldn't matter so much, as the file is loaded on the fly when needed. On the other hand, it would probably be better to have it there just once... 
Does it make sense to extract the data loading outside the DataGenerator class and instead of the path pass something like a pointer to the data? Or do you have a different idea?
